editor/editor_file_system.cpp

```diff
@@ -749,28 +769,50 @@ void EditorFileSystem::scan() {
 		Thread::Settings s;
 		scanning = true;
 		scan_total = 0;
+		emit_signal(SNAME("scan_started"), true);
 		s.priority = Thread::PRIORITY_LOW;
 		thread.start(_thread_func, this, s);
-		//tree->hide();
-		//progress->show();
 	}
 }
 
-void EditorFileSystem::ScanProgress::update(int p_current, int p_total) const {
-	float ratio = low + ((hi - low) / p_total) * p_current;
-	progress->step(ratio * 1000);
-	EditorFileSystem::singleton->scan_total = ratio;
+void EditorFileSystem::ScanProgress::increment() {
+	real++;
+	nb_file_to_scan--;
+	current = CLAMP(real / estimated, current, 1);
+	progress->step(current * 1000);
+	EditorFileSystem::singleton->scan_total = current;
 }
 
-EditorFileSystem::ScanProgress EditorFileSystem::ScanProgress::get_sub(int p_current, int p_total) const {
-	ScanProgress sp = *this;
-	float slice = (sp.hi - sp.low) / p_total;
-	sp.low += slice * p_current;
-	sp.hi = slice;
-	return sp;
+void EditorFileSystem::ScanProgress::update_estimated(int p_depth, int p_nb_sub_dirs, int p_nb_files) {
+	// The algorithm calculates an estimated value for the progress of scanning directories and files.
+	// It adjusts the estimate based on the number of directories and files discovered so far,
+	// and ensures the estimated value does not decrease drastically between updates.
+	float average = 10;
+	directories_processed++;
+	if (nb_dir_to_scan > 0) {
+		nb_dir_to_scan--;
+	}
+	nb_dir_to_scan += p_nb_sub_dirs;
+	nb_file_to_scan += p_nb_files;
+
+	if (directories_processed > 2) {
+		average = (real + nb_file_to_scan) / directories_processed;
+		if (average < 10) {
+			average = 10;
+		}
+	}
+
+	float old_estimated = estimated;
+	estimated = real + nb_file_to_scan + (nb_dir_to_scan * average * p_depth);
+	if (estimated < old_estimated * 0.95) {
+		estimated = old_estimated * 0.95;
+	}
+	if (estimated <= real + average) {
+		estimated = real + average;
+	}
```

Comment:
```suggestion
	float min_estimated = MAX(old_estimated * 0.95f, real + average);
	float new_estimated = real + nb_file_to_scan + (nb_dir_to_scan * average * p_depth);
	estimated = MAX(min_estimated, new_estimated);
```

Please double-check the logic is equivalent, though.

Reply:
Yes, it seems to be equivalent. Done!